`OperatorWrapCheck` throws NPE on guarded patterns

Link to check documentation: https://checkstyle.sourceforge.io/config_whitespace.html#OperatorWrap

Noticed in https://github.com/checkstyle/checkstyle/pull/11186  at https://github.com/checkstyle/checkstyle/runs/4785736479?check_suite_focus=true :



```

➜  javac --enable-preview --release 17 Test.java             
Note: Test.java uses preview features of Java SE 17.
Note: Recompile with -Xlint:preview for details.

➜  cat Test.java                                             
public class Test {
    String typeGuardAfterParenthesizedTrueSwitchStatement(Object o) {
        switch (o) {
            case (Integer i) && i == 0: o = String.valueOf(i); return "true";
            case ((Integer i) && i == 2): o = String.valueOf(i); return "second";
            case Object x: return "any";
        }
    }

    String typeGuardAfterParenthesizedTrueSwitchExpression(Object o) {
        return switch (o) {
            case (Integer i) && i == 0: o = String.valueOf(i); yield "true";
            case ((Integer i) && i == 2): o = String.valueOf(i); yield "second";
            case Object x: yield "any";
        };
    }

    String typeGuardAfterParenthesizedTrueIfStatement(Object o) {
        if (o != null && o instanceof ((Integer i) && i == 0)) {
            return "true";
        } else if (o != null && o instanceof (((Integer i) && i == 2)) && (o = i) != null) {
            return "second";
        } else {
            return "any";
        }
    }
}

➜  cat config.xml                                            
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="OperatorWrap">
        </module>
    </module>
</module>

➜  java -jar checkstyle-9.3-SNAPSHOT-all.jar -c config.xml Test.java
Starting audit...
com.puppycrawl.tools.checkstyle.api.CheckstyleException: Exception was thrown while processing Test.java
        at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:305)
        at com.puppycrawl.tools.checkstyle.Checker.process(Checker.java:224)
        at com.puppycrawl.tools.checkstyle.Main.runCheckstyle(Main.java:407)
        at com.puppycrawl.tools.checkstyle.Main.runCli(Main.java:330)
        at com.puppycrawl.tools.checkstyle.Main.execute(Main.java:189)
        at com.puppycrawl.tools.checkstyle.Main.main(Main.java:126)
Caused by: java.lang.NullPointerException: Cannot invoke "com.puppycrawl.tools.checkstyle.api.DetailAST.getType()" because "result" is null
        at com.puppycrawl.tools.checkstyle.checks.whitespace.OperatorWrapCheck.adjustParens(OperatorWrapCheck.java:452)
        at com.puppycrawl.tools.checkstyle.checks.whitespace.OperatorWrapCheck.getLeftNode(OperatorWrapCheck.java:394)
        at com.puppycrawl.tools.checkstyle.checks.whitespace.OperatorWrapCheck.isNewLineModeViolation(OperatorWrapCheck.java:345)
        at com.puppycrawl.tools.checkstyle.checks.whitespace.OperatorWrapCheck.visitToken(OperatorWrapCheck.java:308)
        at com.puppycrawl.tools.checkstyle.TreeWalker.notifyVisit(TreeWalker.java:335)
        at com.puppycrawl.tools.checkstyle.TreeWalker.processIter(TreeWalker.java:406)
        at com.puppycrawl.tools.checkstyle.TreeWalker.walk(TreeWalker.java:273)
        at com.puppycrawl.tools.checkstyle.TreeWalker.processFiltered(TreeWalker.java:154)
        at com.puppycrawl.tools.checkstyle.api.AbstractFileSetCheck.process(AbstractFileSetCheck.java:87)
        at com.puppycrawl.tools.checkstyle.Checker.processFile(Checker.java:331)
        at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:292)
        ... 5 more
Checkstyle ends with 1 errors.


```

I would expect `OperatorWrapCheck` to process `Test.java` without throwing a null pointer exception.

We can remove file filter for this file from `openjdk17-excluded.files` in PR for this issue, and exclusion from any project files in contribution.

--- a/.ci/openjdk17-excluded.files
+++ b/.ci/openjdk17-excluded.files
@@ -29,11 +29,6 @@
   <property name="fileNamePattern" value="[\\/]test[\\/]langtools[\\/]tools[\\/]javac[\\/]unicode[\\/]FirstChar2.java$"/>
 </module>
 
-<!-- until https://github.com/checkstyle/checkstyle/issues/11187-->
-<module name="BeforeExecutionExclusionFileFilter">
-  <property name="fileNamePattern" value="[\\/]test[\\/]langtools[\\/]tools[\\/]javac[\\/]patterns[\\/]Guards.java$"/>
-</module>
-
 <!-- non-compilable -->
 <module name="BeforeExecutionExclusionFileFilter">
   <property name="fileNamePattern" value="[\\/]test[\\/]langtools[\\/]tools[\\/]javac[\\/]unicode[\\/]SupplementaryJavaID1.java$"/>

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java
@@ -390,6 +390,9 @@ public class OperatorWrapCheck
         if (node.getFirstChild() == null || isAssignToVariable(node)) {
             result = node.getPreviousSibling();
         }
+        else if (isInPatternDefinition(node)) {
+            result = node.getFirstChild();
+        }
         else {
             result = adjustParens(node.getFirstChild(), DetailAST::getNextSibling);
         }
@@ -397,6 +400,30 @@ public class OperatorWrapCheck
             result = result.getLastChild();
         }
         return result;
+    }
+
+    /**
+     * Ascends AST to determine if given node is part of a pattern
+     * definition.
+     *
+     * @param node the node to check
+     * @return true if node is in pattern definition
+     */
+    private static boolean isInPatternDefinition(DetailAST node) {
+        DetailAST parent = node;
+        final int[] tokensToStopOn = {
+            // token we are looking for
+            TokenTypes.PATTERN_DEF,
+            // tokens that mean we can stop looking
+            TokenTypes.EXPR,
+            TokenTypes.RESOURCE,
+            TokenTypes.COMPILATION_UNIT,
+        };
+
+        do {
+            parent = parent.getParent();
+        } while (!TokenUtil.isOfType(parent, tokensToStopOn));
+        return parent.getType() == TokenTypes.PATTERN_DEF;
     }
 
     /**

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheckTest.java
@@ -139,6 +139,31 @@ public class OperatorWrapCheckTest
     }
 
     @Test
+    public void testGuardedPattern() throws Exception {
+        final String[] expected = {
+            "52:30: " + getCheckMessage(MSG_LINE_NEW, "&&"),
+            "54:31: " + getCheckMessage(MSG_LINE_NEW, "&&"),
+            "62:30: " + getCheckMessage(MSG_LINE_NEW, "&&"),
+            "64:31: " + getCheckMessage(MSG_LINE_NEW, "&&"),
+            "71:23: " + getCheckMessage(MSG_LINE_NEW, "&&"),
+            "72:43: " + getCheckMessage(MSG_LINE_NEW, "&&"),
+            "75:30: " + getCheckMessage(MSG_LINE_NEW, "&&"),
+            "76:44: " + getCheckMessage(MSG_LINE_NEW, "&&"),
+            "77:34: " + getCheckMessage(MSG_LINE_NEW, "&&"),
+        };
+
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputOperatorWrapGuardedPatterns.java"), expected);
+    }
+
+    @Test
+    public void testTryWithResources() throws Exception {
+        final String[] expected = CommonUtil.EMPTY_STRING_ARRAY;
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputOperatorWrapTryWithResources.java"), expected);
+    }
+
+    @Test
     public void testInvalidOption() throws Exception {
 
         try {

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/operatorwrap/InputOperatorWrapGuardedPatterns.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/operatorwrap/InputOperatorWrapGuardedPatterns.java
@@ -1,0 +1,84 @@
+/*
+OperatorWrap
+option = (default)nl
+tokens = (default)QUESTION, COLON, EQUAL, NOT_EQUAL, DIV, PLUS, MINUS, STAR, MOD, \
+         SR, BSR, GE, GT, SL, LE, LT, BXOR, BOR, LOR, BAND, LAND, TYPE_EXTENSION_AND, \
+         LITERAL_INSTANCEOF
+
+
+*/
+//non-compiled with javac: Compilable with Java17
+
+package com.puppycrawl.tools.checkstyle.checks.whitespace.operatorwrap;
+
+public class InputOperatorWrapGuardedPatterns {
+    String typeGuardAfterParenthesizedTrueSwitchStatement1(Object o) {
+        switch (o) {
+            case (Integer i)
+                    && i == 0: o = String.valueOf(i); return "true"; // ok
+            case ((Integer i)
+                    && i == 2): o = String.valueOf(i); return "second"; // ok
+            case Object x: return "any";
+        }
+    }
+
+    String typeGuardAfterParenthesizedTrueSwitchExpression1(Object o) {
+        return switch (o) {
+            case (Integer i)
+                    && i == 0: o = String.valueOf(i); yield "true"; // ok
+            case ((Integer i)
+                    && i == 2): o = String.valueOf(i); yield "second"; // ok
+            case Object x: yield "any";
+        };
+    }
+
+    String typeGuardAfterParenthesizedTrueIfStatement1(Object o) {
+        if (o != null
+                && o instanceof ((Integer i) // ok
+                && i == 0)) { // ok
+            return "true";
+        } else if (o != null
+                && o instanceof (((Integer i)  // ok
+                && i == 2))  // ok
+                && (o = i) != null) { // ok
+            return "second";
+        } else {
+            return "any";
+        }
+    }
+
+    String typeGuardAfterParenthesizedTrueSwitchStatement2(Object o) {
+        switch (o) {
+            case (Integer i) && // violation
+                    i == 0: o = String.valueOf(i); return "true";
+            case ((Integer i) && // violation
+                    i == 2): o = String.valueOf(i); return "second";
+            case Object x: return "any";
+        }
+    }
+
+    String typeGuardAfterParenthesizedTrueSwitchExpression2(Object o) {
+        return switch (o) {
+            case (Integer i) && // violation
+                    i == 0: o = String.valueOf(i); yield "true";
+            case ((Integer i) && // violation
+                    i == 2): o = String.valueOf(i); yield "second";
+            case Object x: yield "any";
+        };
+    }
+
+    String typeGuardAfterParenthesizedTrueIfStatement2(Object o) {
+        if (o != null && // violation
+                o instanceof ((Integer i) && // violation
+                        i == 0)) {
+            return "true";
+        } else if (o != null && // violation
+                o instanceof (((Integer i) && // violation
+                        i == 2)) && // violation
+                (o = i) != null) {
+            return "second";
+        } else {
+            return "any";
+        }
+    }
+}

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/operatorwrap/InputOperatorWrapTryWithResources.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/operatorwrap/InputOperatorWrapTryWithResources.java
@@ -1,0 +1,54 @@
+/*
+OperatorWrap
+option = (default)nl
+tokens = (default)QUESTION, COLON, EQUAL, NOT_EQUAL, DIV, PLUS, MINUS, STAR, MOD, \
+         SR, BSR, GE, GT, SL, LE, LT, BXOR, BOR, LOR, BAND, LAND, TYPE_EXTENSION_AND, \
+         LITERAL_INSTANCEOF
+
+
+*/
+//non-compiled with javac: Compilable with Java9
+package com.puppycrawl.tools.checkstyle.checks.whitespace.operatorwrap;
+
+public class InputOperatorWrapTryWithResources implements AutoCloseable
+{
+    public static void main(String[] args) throws Exception {
+        InputOperatorWrapTryWithResources v = new InputOperatorWrapTryWithResources();
+
+        try (v.finalWrapper.finalField) {
+        }
+
+        try (new InputOperatorWrapTryWithResources() { }.finalWrapper.finalField) {
+        }
+
+        try ((args.length > 0 ? v
+                : new InputOperatorWrapTryWithResources()).finalWrapper.finalField) { // ok
+        }
+
+        //More than one resource
+        InputOperatorWrapTryWithResources i1 = new InputOperatorWrapTryWithResources();
+        try (i1; InputOperatorWrapTryWithResources i2 = new InputOperatorWrapTryWithResources()) {
+        }
+
+        InputOperatorWrapTryWithResources m1 = new InputOperatorWrapTryWithResources();
+        try (m1; InputOperatorWrapTryWithResources m2 = m1; // ok
+             InputOperatorWrapTryWithResources m3 = m2;) {
+        }
+    }
+
+    final static FinalWrapper finalWrapper = new FinalWrapper();
+    public void method() throws Exception {
+        try(this.finalWrapper.finalField) {
+        }
+    }
+    static class FinalWrapper {
+        public final AutoCloseable finalField = new AutoCloseable() {
+            @Override
+            public void close() throws Exception {
+            }
+        };
+    }
+    @Override
+    public void close() throws Exception {
+    }
+}

